### PR TITLE
Refactored decider to rely on initiatedEventId instead of string ids

### PR DESCRIPTION
**Pull request by mfateev**
The previous decider implementation used activityId, childWorkflowId and signalId to match events to appropriate state machines. This approach is error prone and makes certain code paths complicated. 
The correct way is to use the initiatedEventId to correlate events. This wasn't possible before as server didn't provide guarantees that every decision has only one correspondent event in a history.
Now server supports this guarantee and this PR implements the migration.

It also includes refactoring of the main ReplayDecider loop as well as support for SideEffect.

### CODE REVIEW

File: src/main/java/com/uber/cadence/internal/testservice/TestWorkflowMutableStateImpl.java [hunk 1]
```diff
@@ -576,70 +599,68 @@ public void childWorkflowStarted(ChildWorkflowExecutionStartedEventAttributes a)
...
+      String activityId, ChildWorkflowExecutionFailedEventAttributes a, long initiatedEventId)
       throws InternalServiceError, EntityNotExistsError, BadRequestError {
     update(
         ctx -> {
           String childId = a.getWorkflowExecution().getWorkflowId();
-          StateMachine<ChildWorkflowData> child = getChildWorkflow(childId);
+          StateMachine<ChildWorkflowData> child = getChildWorkflow(initiatedEventId);
           child.action(StateMachines.Action.FAIL, ctx, a, 0);
-          childWorkflows.remove(childId);
+          childWorkflows.remove(a.getInitiatedEventId());
```
Comment (meiliang86) [inline src/main/java/com/uber/cadence/internal/testservice/TestWorkflowMutableStateImpl.java]: Seems like attribute has eventID in it. Can we just always use that and avoid the interface change?

File: src/main/java/com/uber/cadence/internal/replay/ClockDecisionContext.java [hunk 2]
```diff
@@ -122,4 +132,34 @@ private void timerCancelled(String timerId, Exception reason) {
...
+    } else {
+      try {
+        result = func.apply();
+      } catch (Error e) {
+        throw e;
+      } catch (Exception e) {
+        throw new Error("sideEffect function failed", e);
+      }
+    }
+    decisions.recordMarker(SIDE_EFFECT_MARKER_NAME, result);
```
[the remark this answers is not in the document]
Comment (mfateev) [reply src/main/java/com/uber/cadence/internal/replay/ClockDecisionContext.java]: Yes. As we are using eventId as a decision id we have to add all decisions even on replay to ensure that ids match.

File: src/main/java/com/uber/cadence/internal/replay/HistoryHelper.java [hunk 3]
```diff
@@ -17,102 +17,250 @@
...
+  /**
+   * Iterates through decisions and returns one DecisionEvents instance per DecisionTaskStarted
+   * event.
+   */
+  static class DecisionEventsIterator implements Iterator<DecisionEvents> {
 
-    public boolean isNextDecisionFailed() {
+    private EventsIterator events;
+
+    private Queue<HistoryEvent> bufferedEvents = new ArrayDeque<>();
```
[the remark this answers is not in the document]
Comment (mfateev) [reply src/main/java/com/uber/cadence/internal/replay/HistoryHelper.java]: Good catch. It is leftover from SWF which doesn't implement server side event buffering as Cadence does.

File: src/main/java/com/uber/cadence/internal/testservice/TestWorkflowMutableStateImpl.java (hunk 1 above)
Comment (mfateev) [reply src/main/java/com/uber/cadence/internal/testservice/TestWorkflowMutableStateImpl.java]: Good point. It looks like it was leftover from the original version of the refactoring.

**Pull request merged**